provider/secret_descriptor.go

```diff
@@ -337,29 +345,70 @@ func NewSecretDescriptorList(mountDir, translate, objectSpec string, regions []s
+		// | 1 | duplicate    | empty                |
+		// | 2 | empty        | duplicate            |
+		// | 3 | duplicate    | duplicate            |
+		// | 4 | empty        | empty                |
+		// -------------------------------------------
+		found, ok := seenNames[descriptor.ObjectName]
+		if ok {
+			descHasAlias := descriptor.ObjectAlias != ""
+			foundHasAlias := found.ObjectAlias != ""
+			descHasVersionLabel := descriptor.ObjectVersionLabel != ""
+			foundHasVersionLabel := found.ObjectVersionLabel != ""
+
+			if descHasAlias && foundHasAlias {
+				if descriptor.ObjectAlias == found.ObjectAlias {
+					if descHasVersionLabel && foundHasVersionLabel {
+						if descriptor.ObjectVersionLabel == found.ObjectVersionLabel {
+							// Case 3
+							return nil, fmt.Errorf("found descriptor with duplicate object name %s, object alias %s, and version label %s", descriptor.ObjectName, descriptor.ObjectAlias, descriptor.ObjectVersionLabel)
+						}
+					} else {
+						// Case 1
+						return nil, fmt.Errorf("found descriptor with duplicate object name %s and object alias %s with no version label", descriptor.ObjectName, descriptor.ObjectAlias)
+					}
+				}
+			} else {
+				if descHasVersionLabel && foundHasVersionLabel {
+					if descriptor.ObjectVersionLabel == found.ObjectVersionLabel {
+						// Case 2
+						return nil, fmt.Errorf("found descriptor with duplicate object name %s and version label %s with no object alias", descriptor.ObjectName, descriptor.ObjectVersionLabel)
+					}
+				} else {
+					// Case 4
+					return nil, fmt.Errorf("found descriptor with duplicate object name %s with no object alias or version label", descriptor.ObjectName)
+				}
+			}
+		}
+		// Add the descriptor to the seenNames map after validation
+		seenNames[descriptor.ObjectName] = SecretDescriptorSlice{
+			ObjectAlias:        descriptor.ObjectAlias,
+			ObjectVersionLabel: descriptor.ObjectVersionLabel,
 		}
-		names[descriptor.ObjectName] = true
 
-		if len(descriptor.ObjectAlias) > 0 {
-			if names[descriptor.ObjectAlias] {
-				return nil, fmt.Errorf("Name already in use for objectAlias: %s", descriptor.ObjectAlias)
-			}
-			names[descriptor.ObjectAlias] = true
+		if seenAliases[descriptor.ObjectAlias] {
+			return nil, fmt.Errorf("found duplicate object alias %s", descriptor.ObjectAlias)
+		}
+		// Add the object alias to the seenAliases map for use in JMES path validation below
+		if descriptor.ObjectAlias != "" {
+			seenAliases[descriptor.ObjectAlias] = true
 		}
 
 		if len(descriptor.JMESPath) == 0 { //jmesPath not used. No more checks
 			continue
 		}
 
 		for _, jmesPathEntry := range descriptor.JMESPath {
-			if names[jmesPathEntry.ObjectAlias] {
-				return nil, fmt.Errorf("Name already in use for objectAlias: %s", jmesPathEntry.ObjectAlias)
+			if seenAliases[jmesPathEntry.ObjectAlias] {
+				return nil, fmt.Errorf("found duplicate object alias %s in JMES path entry %s", jmesPathEntry.ObjectAlias, jmesPathEntry.Path)
```

Comment:
I'm a bit concerned about this breaking customers that search for this error message, should we keep the start of the string the same?